### PR TITLE
Open URL through automate for various objects

### PR DESCRIPTION
**Pull request by martinpovolny**
Adding a table for temporary storing URL returned from automate.

BZ:
https://bugzilla.redhat.com/show_bug.cgi?id=1550002

UI PR: https://github.com/ManageIQ/manageiq-ui-classic/pull/5688
Guides article: https://github.com/martinpovolny/guides/blob/automate_url_opener/automate_url_open.md (from PR: https://github.com/ManageIQ/guides/pull/355)
Core PR: https://github.com/ManageIQ/manageiq/pull/18849
Automate PR: https://github.com/ManageIQ/manageiq-automation_engine/pull/328

### CODE REVIEW

**Comment by martinpovolny** [comment]
ping @Ladas

File: db/migrate/20190604090631_create_saved_urls.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class CreateSavedUrls < ActiveRecord::Migration[5.0]
+  def change
+    create_table :saved_urls do |t|
+      t.bigint  "refers_to_id", :type => :bigint
+      t.string  "refers_to_type"
```
Comment (Ladas) [inline db/migrate/20190604090631_create_saved_urls.rb]: `t.references :refers_to, polymorphic: true, index: true`

should create the `refers_to_id` and ..._type` and the index

File: db/migrate/20190604090631_create_saved_urls.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class CreateSavedUrls < ActiveRecord::Migration[5.0]
+  def change
+    create_table :saved_urls do |t|
```
Comment (Ladas) [inline db/migrate/20190604090631_create_saved_urls.rb]: should be `create_table :saved_urls, :id => :bigserial` I believe

**Comment by Ladas** [comment]
cc @lpichler what's the rake command to generate the schema? We should still run it, right? (The CI was failing without it in the past)

**Comment by martinpovolny** [comment]
@Ladas : PR updated according to your guidance. Please, check again.

File: db/migrate/20190604090631_create_external_urls.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class CreateExternalUrls < ActiveRecord::Migration[5.0]
+  def change
+    create_table :external_urls, :id => :bigserial do |t|
+      t.references :refers_to, :polymorphic => true, :index => true, :type => :bigint
```
Comment (Ladas) [inline db/migrate/20190604090631_create_external_urls.rb]: if the `index => true` here does't work, lets remove it (it should create `external_urls` for you, but it's not implemented in all migration methods)

File: db/migrate/20190604090631_create_external_urls.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+      t.string  "url"
```
Comment (lpichler) [inline db/migrate/20190604090631_create_external_urls.rb]: nitpick: but we are using symbols:

`t.string  "url"`

File: db/migrate/20190604090631_create_external_urls.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+      t.belongs_to :user, :type => :bigint
+    end
+
+    add_index "external_urls", ["refers_to_id", "refers_to_type"], :name => "saved_urls_on_refers_to_id_and_refers_to_type"
```
Comment (lpichler) [inline db/migrate/20190604090631_create_external_urls.rb]: same here

**Comment by lpichler** [comment]
> cc @lpichler what's the rake command to generate the schema? We should still run it, right? (The CI was failing without it in the past)

looks like it was removed: http://talk.manageiq.org/t/new-schema-specs-for-new-replication/1404/4

File: db/migrate/20190604090631_create_external_urls.rb (hunk 3 above)
Comment (lpichler) [reply db/migrate/20190604090631_create_external_urls.rb]: `refers_to` maybe just `resource` ?

**Comment by martinpovolny** [comment]
@Ladas : fixed.

@lpichler : renamed and also renamed in the core PR here: https://github.com/ManageIQ/manageiq/pull/18849

**Comment by lpichler** [comment]
@miq-bot assign @carbonin

File: db/migrate/20190604090631_create_external_urls.rb (hunk 4 above)
Comment (carbonin) [reply db/migrate/20190604090631_create_external_urls.rb]: Agree with @lpichler here, can we use all symbols for the column names?

**Comment by carbonin** [comment]
@martinpovolny I think the BZ link should be https://bugzilla.redhat.com/show_bug.cgi?id=1643331, right?

The one in the initial PR comment is closed.
Also could you add it to the commit message so the bot will pick it up and comment on the BZ?

**Comment by carbonin** [comment]
Also, what objects are going to be referenced by this url and if it's temporary, why do we need a table?

**Comment by agrare** [comment]
@martinpovolny If the goal is to show the vm summary page why a new table and not an attribute on the vm like we do with topology?

Also, the URL for vsphere is going to be different depending on version, if you have the html5 UI or just the old flash UI, etc... so I'm curious who would be generating this URL.

E.g. for vSphere 6.7 it looks like this `https://dev-vc67/ui/#?extensionId=vsphere.core.inventory.serverObjectViewsExtension&objectId=urn:vmomi:VirtualMachine:vm-337:a6b3087a-3d6b-4852-9732-17094d702269&navigator=vsphere.core.viTree.vmsAndTemplatesView`

**Comment by martinpovolny** [comment]
@carbonin : the link is right.

> @martinpovolny If the goal is to show the vm summary page why a new table and not an attribute on the vm like we do with topology?

No that is not the goal. The goal is in the BZ. If the goal is not clear from the BZ, then, please, read the (partial) documentation in this PR: https://github.com/ManageIQ/guides/pull/355

This PR has nothing to do with the VM consoles although it can be used to replace a VM console with a custom one.

**Comment by martinpovolny** [comment]
> Also, what objects are going to be referenced by this url and if it's temporary, why do we need a table?

How else do we pass data from a worker running on one appliance to a different appliance?

**Comment by agrare** [comment]
The bz you have linked to in the description is closed as a duplicate of https://bugzilla.redhat.com/show_bug.cgi?id=1643331

The goal of the BZ is pretty clear:
> Description of request:
> 
> Ability to create CloudForms custom buttons which can open an HTTPS URL to a provider's web client.

Also important that this also isn't the VM's console, but the VM's summary page.

File: db/migrate/20190604090631_create_external_urls.rb (hunk 4 above)
Comment (martinpovolny) [reply db/migrate/20190604090631_create_external_urls.rb]: I missed that. Fixed now.

**Comment by martinpovolny** [comment]
@NickLaMuro, @agrare: this is about extending an existing mechanism (mentioned for example here:
http://talk.manageiq.org/t/lauch-ssh-error/4343/15 ) to work with several new types of entities.

I have implemented that mechanism several years ago, and at that time I did put the URL in the `SystemConsole` table. That limits the usefulnes of that to just VMs.

Here I am trying to do the same in a more universal way.

I do not know how else to explain it to you. Above I have linked a PR to a guides PR where the mechanism is explained: https://github.com/martinpovolny/guides/blob/automate_url_opener/automate_url_open.md.

Do you need any further explanation? Shall we schedule a video chat?

**Comment by NickLaMuro** [comment]
@martinpovolny Wrong "Nick" I think...

cc @carbonin

**Comment by martinpovolny** [comment]
> The bz you have linked to in the description is closed as a duplicate of > https://bugzilla.redhat.com/show_bug.cgi?id=1643331

I have linked the one that contains a better explanation of the issue. On purpose.

You are linking the one that only mentiones one usecase and very shortly. Please, read the BZ that I linked.

**Comment by agrare** [comment]
Okay sounds like this bz https://bugzilla.redhat.com/show_bug.cgi?id=1643331 should have actually been marked as a duplicate of https://bugzilla.redhat.com/show_bug.cgi?id=1550002 then?  We shouldn't be pointing PRs to closed BZs so lets either reopen the original one or add enough description to the current BZ.

**Comment by martinpovolny** [comment]
I am glad we moved forward ;-)

Are there any changes needed here on top of what @Ladas and @lpichler requested (and I already did)?

**Pull request merged**